Edits to clarify usage of pg_waldump on TDE-enabled cluster (#5904)
EnterpriseDB/docs: product_docs/docs/tde/15/troubleshooting.mdx

---
title: "Troubleshooting with encrypted WAL files"
navTitle: WAL files
deepToC: true
---

When TDE is enabled, WAL files are encrypted. If you want to perform operations on the encrypted WAL files, you will need to allow the operation to decrypt the file. 

When troubleshooting with encrypted WAL files, you can use WAL command options. 

## Dumping a TDE-encrypted WAL file

> To work with an encrypted WAL file, you need to ensure the [pg_waldump](https://www.postgresql.org/docs/15/pgwaldump.html) utility can access the unwrap key and decrypt it. For this purpose, the utility requires awarenes of three values.  

Reviewer: awarenes > awareness

Reviewer: https://www.postgresql.org/docs/15/pgwaldump.html

This URL references PG15, we should be using the current ver, so it should be https://www.postgresql.org/docs/current/pgwaldump.html

> To work with an encrypted WAL file, you need to ensure the [pg_waldump](https://www.postgresql.org/docs/current/pgwaldump.html) utility can access the unwrap key and decrypt it. For this purpose, the utility requires awareness of three values.  

Reviewer 2: ```suggestion
To work with an encrypted WAL file, you need to ensure the [pg_waldump](https://www.postgresql.org/docs/current/pgwaldump.html) utility can access the unwrap key and use it to decrypt the WAL file. For this purpose, the utility requires three settings, a flag to enable encryption and two values, for the key and for the unwrap command to be used.  
```

> To work with an encrypted WAL file, you need to ensure the [pg_waldump](https://www.postgresql.org/docs/current/pgwaldump.html) utility can access the unwrap key and decrypt it. For this purpose, the utility requires awareness of three values.  

Author: ```suggestion
To work with an encrypted WAL file, you need to ensure the [pg_waldump](https://www.postgresql.org/docs/current/pgwaldump.html) utility can access the unwrap key and decrypt it. For this purpose, the utility requires awareness of three values that you set when initializing the TDE-enabled cluster. 
```

> Pass these values using the following options to the `pg_waldump` command.

Author: ```suggestion
Pass these values using the following options to the `pg_waldump` command. Ensure you use the exact same values you used when initializing the TDE-enabled cluster. 
```

Author: @ebgitelman 
The gist of it is: 
When a customer creates a database cluster with TDE, they have to specify certain values with command line flags/options. 

Later, if they need to do any operations on WAL files, as this page is doing, they will need to "remember" the flags they used in the first place when initializing the cluster, and then place them here again for decryption. I have tried to make that clear in each of the three sections, and now I have also added this sentence here. 

Do you think it is clear?

Reviewer 3: Yes, looks good (with my edit to steamline)

### `--data-encryption`

Specify this option if the WAL files were encrypted by transparent data encryption.

The `--data-encryption` or `-y` option ensures the command is aware of the encryption. Otherwise, `pg_waldump` can't detect whether WAL files are encrypted.

> It considers the WAL files to encrypt and decrypts them before processing them. Optionally, specify an AES key length. Valid values are `128` and `256`. The default is `128`.

Reviewer 3: I couldn't make out what this means. What does "consider" mean in this context. Does it evaluate them for some purpose?

Author: Thanks for pointing this out, as I had forgotten to rephrase this part. WDYT? 

```suggestion
Provide the same encryption configuration you used when initializing the TDE-enabled database cluster. For example, if you specified an AES key length during the cluster creation, you must specify it here as well. Otherwise, run the flag with no values.
```

> It considers the WAL files to encrypt, and decrypts them before processing them. Optionally, specify an AES key length. Valid values are 128 and 256. The default is 128.

Reviewer 2: Not sure where this para thinks it should be but not here. --data-encryption takes no parameters.

> It considers the WAL files to encrypt, and decrypts them before processing them. Optionally, specify an AES key length. Valid values are 128 and 256. The default is 128.

Author: Specify that this is only valid/required, if you initialized the cluster with a customized AES key length. If not, the flag with no value is enough. 

Also, I could link to https://www.enterprisedb.com/docs/tde/latest/enabling_tde/#using-initdb-tde-options for more info.

### `--key-file-name=<file>`

Use the `--key-file-name=<file>` option to reference the data encryption key required to decrypt the WAL file. Provide the location of the `pg_encryption/key.bin` file. This file is generated when you initialize a cluster with encryption enabled. 

The command can then load the data encryption key from the given location. 

### `--key-unwrap-command=<command>`

For the `--key-unwrap-command=<command>` option, provide the decryption command you specified to unwrap (decrypt) the data encryption key when initializing the TDE cluster. See [Using initdb TDE options](enabling_tde/#using-initdb-tde-options) for more information. 

> Alternatively, you can set the `PGDATAKEYUNWRAPCMD` environment variable before running the `pg_waldump` command. `pg_waldump` automatically falls back on `PGDATAKEYUNWRAPCMD`, if the `--key-unwrap-command=<command>` options is not specified. You can see how to export an environment variable in this [cluster intialization example](enabling_tde/#example).

Reviewer: intialization > initialization

### Example

This example uses `pg_waldump` to display the WAL log of an encrypted cluster that uses `openssl` to wrap the data encryption key: 

```
pg_waldump --data-encryption --key-file-name=pg_encryption/key.bin --key-unwrap-command='openssl enc -d -aes-128-cbc -pass pass:<passphrase> -in %p'
```

## Resetting a corrupt TDE-encrypted WAL file

> To reset a corrupt encrypted WAL file, you must ensure the [pg_resetwal](https://www.postgresql.org/docs/15/app-pgresetwal.html) command can access the unwrap key and decrypt it. You can either pass the key for the unwrap command using the following option to the `pg_resetwal` command or depend on the fallback environment variable:

Reviewer: https://www.postgresql.org/docs/15/app-pgresetwal.html

This URL references PG15, we should be using the current ver, so it should be https://www.postgresql.org/docs/current/app-pgresetwal.html

> To reset a corrupt encrypted WAL file, you must ensure the [pg_resetwal](https://www.postgresql.org/docs/current/app-pgresetwal.html) command can access the unwrap key and decrypt it. You can either pass the key for the unwrap command using the following option to the `pg_resetwal` command or depend on the fallback environment variable:

Reviewer 2: ```suggestion
To reset a corrupt encrypted WAL file, you must ensure the [pg_resetwal](https://www.postgresql.org/docs/current/app-pgresetwal.html) command can access the unwrap key and use it to decrypt the WAL file. You can either pass the key for the unwrap command using the following option to the `pg_resetwal` command or depend on the fallback environment variable:
```

### `--key-unwrap-command=<command>`

For the `--key-unwrap-command=<command>` option, provide the decryption command you specified to unwrap (decrypt) the data encryption key when initializing the TDE cluster. See [Using initdb TDE options](enabling_tde/#using-initdb-tde-options) for more information. 

> Alternatively, you can set the `PGDATAKEYUNWRAPCMD` environment variable before running the `pg_waldump` command. `pg_waldump` automatically falls back on `PGDATAKEYUNWRAPCMD`, if the `--key-unwrap-command=<command>` options is not specified. You can see how to export an environment variable in this [cluster intialization example](enabling_tde/#example).

Reviewer: The `pg_waldump` here should be `pg_resetwal`
